### PR TITLE
tmpfiles: read /etc/passwd + /etc/group with fgetpwent()/fgetgrent() if --root= is specified

### PR DESCRIPTION
**Pull request by poettering**
Fixes: #14806

### CODE REVIEW

**Comment by nolange** [comment]
Great!

But the [testcase.tar.gz](https://github.com/systemd/systemd/files/4164869/testcase.tar.gz) from the related issue is not working:

```
$ SYSTEMD_LOG_LEVEL=debug fakeroot -u  ./runme.sh 
Looking for configuration files in (higher priority first):
	/etc/tmpfiles.d
	/run/tmpfiles.d
	/usr/local/lib/tmpfiles.d
	/usr/lib/tmpfiles.d
Reading config file "/tmp/testcase/etc/tmpfiles.d/var_test.conf"…
Running create action for entry d /tmp/testcase/var/test/foreignuser
Found existing directory "/tmp/testcase/var/test/foreignuser".
"/tmp/testcase/var/test/foreignuser" matches temporary mode 755 already.
Changing "/tmp/testcase/var/test/foreignuser" to owner 5999:4294967295
fchownat() of /tmp/testcase/var/test/foreignuser failed: No such file or directory
Running create action for entry d /tmp/testcase/var/test/foreigngroup
Found existing directory "/tmp/testcase/var/test/foreigngroup".
"/tmp/testcase/var/test/foreigngroup" matches temporary mode 755 already.
Changing "/tmp/testcase/var/test/foreigngroup" to owner 4294967295:5998
fchownat() of /tmp/testcase/var/test/foreigngroup failed: No such file or directory
foreigngroup should be 0 5998:1000 1000
foreignuser  should be 5999 0:1000 1000
```

Whats a bit weird is, that there is a different error, when `var/test` directory is missing:

```
$ SYSTEMD_LOG_LEVEL=debug fakeroot -u  ./runme.sh 
Looking for configuration files in (higher priority first):
	/etc/tmpfiles.d
	/run/tmpfiles.d
	/usr/local/lib/tmpfiles.d
	/usr/lib/tmpfiles.d
Reading config file "/tmp/testcase/etc/tmpfiles.d/var_test.conf"…
Failed to determine whether '/tmp/testcase/var/test/foreignuser' is below autofs, ignoring: No such file or directory
Running create action for entry d /tmp/testcase/var/test/foreignuser
Detected unsafe path transition /tmp/testcase/var → /tmp/testcase/var/test during canonicalization of /tmp/testcase/var/test.
Failed to determine whether '/tmp/testcase/var/test/foreigngroup' is below autofs, ignoring: No such file or directory
Running create action for entry d /tmp/testcase/var/test/foreigngroup
Detected unsafe path transition /tmp/testcase/var → /tmp/testcase/var/test during canonicalization of /tmp/testcase/var/test.
foreigngroup should be 0 5998:stat: cannot stat '/tmp/testcase/var/test/foreigngroup': No such file or directory
foreignuser  should be 5999 0:stat: cannot stat '/tmp/testcase/var/test/foreignuser': No such file or directory


```

**Comment by nolange** [comment]
First issue is in fakeroot, which does not support AT_EMPTY_PATH. Seems to work after fixing that.

Missing `var/tmp` will still complain about unsafe path conversion, but I believe this to be fakeroot issue aswell (-u flag, thats unfortunately needed in my case).

So looks good to me

File: src/tmpfiles/tmpfiles.c [hunk 1]
```diff
@@ -2487,7 +2487,139 @@ static int patch_var_run(const char *fname, unsigned line, char **path) {
...
+                                continue;
+                        }
+                        if (r < 0)
+                                return r;
+
+                        TAKE_PTR(n);
+                }
+
+                *cache = TAKE_PTR(uid_by_name);
+        }
```
Comment (keszybz) [inline src/tmpfiles/tmpfiles.c:2510]: I'd feel more comfortable if this reading was split out to a function. It'd be good to even put all of this new code (the parts to read and query `passwd` and `group` offline) in a separate file so we can add a test for this in the future.

**Comment by poettering** [comment]
regarding synthesizing: will add support for synthesizing root. But nobody is problematic: the definition of nobody might differ between host and root environment, hence we probably shouldn't apply the host's compiled-in defaults to the image.

**Comment by poettering** [comment]
Force pushed new version. Split out the offline stuff into own .c file with own functions. Also synthesizing root now. ptal!

**Comment by keszybz** [comment]
> regarding synthesizing: will add support for synthesizing root.

Great, thanks.

> But nobody is problematic: the definition of nobody might differ between host and root environment, hence we probably shouldn't apply the host's compiled-in defaults to the image.

I'd argue that this is applying the defaults compiled into *systemd-tmpfiles*, which we are already implicitly doing since systemd-tmpfiles changes behaviour a bit in different versions. But yeah, `root` is the important one, and nobody is probably very rarely used.

**Comment by boucman** [comment]
Most cross-compilation systems already have a specific host-systemd recompiled anyway... split-user means we can't trust the host's systemd and we have to compile our own to run systemctl preset-all.

So trusting the compiled-in name and value for nobody makes sense since it's a custom-built host-systemd with the correct values for the target system.

**Comment by nolange** [comment]
Yes, you can always override the default by adding root/nobody to the target rootfs (and potentially removing those again after running systemd-tmpfiles). This is not really an arguments for/against synthesized users, but if you bootstrap a target system you will either use a fitting tool or fitting config files for the setup.

**Comment by poettering** [comment]
so yeah, i started out sharing this, but then i removed it again because the double indexing sysusers needs was too complex for tmpfiles, and it didn't appear worth it... dunno. maybe merge as is? prep a PR that unifies this afterwards, if its good and clean?

**Pull request merged**